server/testutil/app/app.go

```diff
@@ -0,0 +1,220 @@
+	for _, opt := range opts {
+		if err := opt(p); err != nil {
+			t.Fatal(err)
+		}
+	}
+	if p.cmdPath == "" {
+		t.Fatalf("missing `WithBuildBuddyCmdPath` opt")
+	}
+	if p.configPath == "" {
+		t.Fatalf("missing `WithConfigPath` opt")
+	}
+	dataDir, err := ioutil.TempDir("/tmp", "buildbuddy-test-*")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		os.RemoveAll(dataDir)
+	})
+	// NOTE: No SSL ports are required since the server doesn't have an SSL config by default.
+	app := &App{
+		httpPort:       freePort(t),
+		gRPCPort:       freePort(t),
+		monitoringPort: freePort(t),
+	}
+	cmdArgs := []string{
+		fmt.Sprintf("--config_file=%s", p.configPath),
+		fmt.Sprintf("--port=%d", app.httpPort),
+		fmt.Sprintf("--grpc_port=%d", app.gRPCPort),
+		fmt.Sprintf("--monitoring_port=%d", app.monitoringPort),
+		fmt.Sprintf("--app.build_buddy_url=http://localhost:%d", app.httpPort),
+		"--database.data_source=sqlite3://:memory:",
+		fmt.Sprintf("--storage.disk.root_directory=%s", filepath.Join(dataDir, "storage")),
+		fmt.Sprintf("--cache.disk.root_directory=%s", filepath.Join(dataDir, "cache")),
+	}
+	cmdArgs = append(cmdArgs, p.cmdArgs...)
+	cmd := exec.Command(p.cmdPath, cmdArgs...)
+	cmd.Stdout = &app.stdout
+	cmd.Stderr = &app.stderr
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		cmd.Process.Kill() // ignore errors
+	})
+	go func() {
+		err := cmd.Wait()
+		app.mu.Lock()
+		defer app.mu.Unlock()
+		app.exited = true
+		app.err = err
+	}()
+	err = app.waitForReady()
+	if err != nil {
+		t.Fatal(app.fmtErrorWithLogs(err))
+	}
+	return app
+}
+
+// BESBazelFlags returns the Bazel flags required to upload build logs to the App.
+func (a *App) BESBazelFlags() []string {
+	return []string{
+		fmt.Sprintf("--bes_results_url=http://localhost:%d/invocation/", a.httpPort),
+		fmt.Sprintf("--bes_backend=grpc://localhost:%d", a.gRPCPort),
+	}
+}
+
+// RemoteCacheBazelFlags returns the Bazel flags required to use the App's remote cache.
+func (a *App) RemoteCacheBazelFlags() []string {
+	return []string{
+		fmt.Sprintf("--remote_cache=grpc://localhost:%d", a.gRPCPort),
+	}
+}
+
+// Opt customizes the BuildBuddy app instance.
+type RunOpt func(*appParams) error
+
+// WithBuildBuddyCmdPath sets the path to the BuildBuddy binary.
+func WithBuildBuddyCmdPath(path string) RunOpt {
+	return func(p *appParams) error {
+		cmdPath, err := bazelgo.Runfile(path)
+		if err != nil {
+			return err
+		}
+		p.cmdPath = cmdPath
+		return nil
+	}
+}
+
+// WithConfigPath sets the path to the BuildBuddy server config.
+func WithConfigPath(path string) RunOpt {
+	return func(p *appParams) error {
+		configPath, err := bazelgo.Runfile(path)
+		if err != nil {
+			return err
+		}
+		p.configPath = configPath
+		return nil
+	}
+}
+
+// WithArgs adds extra flags to the BuildBuddy server.
+func WithFlags(args ...string) RunOpt {
+	return func(p *appParams) error {
+		p.cmdArgs = append(p.cmdArgs, args...)
+		return nil
+	}
+}
```

Comment:
I've seen this pattern before used with mixed results :)

I think for overriding options, especially when there can be many, it makes a lot of sense. But if you are setting required arguments using this idiom -- it's clearer to just require them as arguments in your Run() function. Can we go back to that?